### PR TITLE
[C#] Add WriteReason parameter to SingleWriter

### PR DESCRIPTION
**Pull request by TedHartMS**

### CODE REVIEW

File: cs/benchmark/Functions.cs [hunk 1]
```diff
@@ -51,13 +51,7 @@ public bool ConcurrentReader(ref Key key, ref Input input, ref Value value, ref
...
         // Upsert functions
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public void SingleWriter(ref Key key, ref Input input, ref Value src, ref Value dst, ref Output output, ref RecordInfo recordInfo, long address)
-        {
-            dst = src;
-        }
-
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public void CopyWriter(ref Key key, ref Value src, ref Value dst, ref RecordInfo recordInfo, long address)
+        public void SingleWriter(WriteReason reason, ref Key key, ref Input input, ref Value src, ref Value dst, ref Output output, ref RecordInfo recordInfo, long address)
```
Comment (badrishc) [inline cs/benchmark/Functions.cs:54]: I think we want the API to be:

```cs
SingleWriter(ref Key key, ref Input input, ref Value src, ref Value dst, ref Output output, ref RecordInfo recordInfo, long address, WriteReason reason)
```

File: cs/src/core/Index/FASTER/FASTER.cs [hunk 2]
```diff
@@ -54,6 +54,24 @@ public enum ReadFlags
         SkipCopyReads = SkipReadCache | SkipCopyToTail,
     }
 
+    /// <summary>
+    /// The reason a SingleWriter was performed
+    /// </summary>
+    public enum WriteReason
```
Comment (badrishc) [inline cs/src/core/Index/FASTER/FASTER.cs:60]: move to its own file for cleanliness?

Comment (TedHartMS) [reply cs/src/core/Index/FASTER/FASTER.cs]: done

File: cs/benchmark/Functions.cs (hunk 1 above)
Comment (TedHartMS) [reply cs/benchmark/Functions.cs]: done

**Pull request merged**